### PR TITLE
Alexeev parser improvements

### PR DESCRIPTION
**Pull request by alexeev509**
Was added supporting of the:
1) Scientific notation for double;
2) Binary literals;
3) Octal literals;
4) Hex literals;
5) Char literals;

### CODE REVIEW

File: src/main/kotlin/org/jetbrains/research/libsl/visitors/ExpressionVisitor.kt [hunk 1]
```diff
@@ -211,83 +224,49 @@ class ExpressionVisitor(
...
-                    ctx.text.dropLast(2).toUByte(),
-                    "ub",
-                    posGetter.getCtxPosition(context.fileName, ctx)
-                )
-                "s" -> IntegerLiteral(
-                    ctx.text.dropLast(1).toShort(),
-                    "s",
-                    posGetter.getCtxPosition(context.fileName, ctx)
-                )
-                "us" -> UnsignedInt16Literal(
```
Comment (dog-m) [inline src/main/kotlin/org/jetbrains/research/libsl/visitors/ExpressionVisitor.kt:236]: If we limiting allowed post-fixes to just `l` and `L`, then how it is possible to parse (and get an appropriate node in AST) other (now unsupported) kinds of integers? *(like `int16` and `int8`)*

File: src/main/antlr/LibSLLexer.g4 [hunk 2]
```diff
@@ -254,6 +254,43 @@ NULL
...
+
+fragment HexIntegerLiteral: HexNumeral IntegerTypeSuffix?;
+
+fragment OctalIntegerLiteral: OctalNumeral IntegerTypeSuffix?;
+
+fragment BinaryIntegerLiteral: BinaryNumeral IntegerTypeSuffix?;
+
+fragment DecimalNumeral: '0' | NonZeroDigit (Digits?);
+
+fragment IntegerTypeSuffix: [lL];
```
Comment (dog-m) [inline src/main/antlr/LibSLLexer.g4]: May be we should extend this "list" a bit.

File: build.gradle.kts [hunk 3]
```diff
@@ -54,4 +55,11 @@ publishing {
             }
         }
     }
+}
+
+tasks.withType<Jar>() {
+    duplicatesStrategy = DuplicatesStrategy.EXCLUDE
+    configurations["compileClasspath"].forEach { file: File ->
+        from(zipTree(file.absoluteFile))
```
Comment (dog-m) [inline build.gradle.kts]: What is the purpose of these changes here?
Looks like a way of making a "fat JAR" to me.

File: src/main/kotlin/org/jetbrains/research/libsl/visitors/ExpressionVisitor.kt [hunk 4]
```diff
@@ -211,83 +224,49 @@ class ExpressionVisitor(
...
 
+    fun convertBinHexOctToPrimitives(num: String, type: String): Number {
+        val result: Number = if (num.startsWith(HEX_PREFIX)) {
+            if ("i".equals(type)) parseInt(num.drop(2), 16) else parseLong(num.drop(2), 16)
+        } else if (num.startsWith(BIN_PREFIX)) {
+            if ("i".equals(type)) parseInt(num.drop(2), 2) else parseLong(num.drop(2), 2)
+        } else if (num.startsWith(OCT_PREFIX) && num.length > 1) {
+            if ("i".equals(type)) parseInt(num.drop(1), 8) else parseLong(num.drop(1), 8)
+        } else {
+            if ("i".equals(type)) num.toInt() else num.toLong()
```
Comment (dog-m) [inline src/main/kotlin/org/jetbrains/research/libsl/visitors/ExpressionVisitor.kt]: Seems to me that it is possible to make a local variable for `"i".equals(type)`

File: src/test/testdata/expected/lsl/charLiterals.lsl [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+library simple;
+typealias Int = int32;
+typealias Char = char;
+
+automaton A : Int {
+    var ch1: Char = 'A';
+    var ch2: Char = 'a';
+    var ch3: Char = 52;
+    var ch4: Char = '*';
+    var ch5: Char = '';
```
Comment (dog-m) [inline src/test/testdata/expected/lsl/charLiterals.lsl]: This line doesn't seem correct to my eye.

File: src/test/testdata/lsl/charLiterals.lsl [hunk 6]
```diff
@@ -0,0 +1,13 @@
+libsl "1.0.0";
+library simple;
+typealias Int = int32;
+typealias Char = char;
+
+automaton A : Int {
+    var ch1: Char = '\u0041';
+    var ch2: Char = 'a';
+    var ch3: Char = 52;
```
Comment (dog-m) [inline src/test/testdata/lsl/charLiterals.lsl]: I don't think it is necessary here.
This line doesn't look like the value should be parsed as `char` and not as `int32`.

File: src/test/testdata/lsl/charLiterals.lsl [hunk 7]
```diff
@@ -0,0 +1,13 @@
+libsl "1.0.0";
+library simple;
+typealias Int = int32;
+typealias Char = char;
+
+automaton A : Int {
+    var ch1: Char = '\u0041';
```
Comment (dog-m) [inline src/test/testdata/lsl/charLiterals.lsl:7]: I suggest adding an example of a more complex Unicode character here (and seeing the issues that may arise from that).
Like `\u72D0` for example.

File: src/test/testdata/lsl/scientificNotation.lsl [hunk 8]
```diff
@@ -0,0 +1,10 @@
+libsl "1.0.0";
+library simple;
+typealias Int = int32;
+typealias float = float32;
+typealias double = float64;
+automaton A : Int {
+    var d1: double = 4.05E-13;
+    var d2: double = 4.05E13;
+    var d3: double = 4.05E+13;
+}
```
Comment (dog-m) [inline src/test/testdata/lsl/scientificNotation.lsl]: I suggest adding negatives, a zero and numbers with suffixes.
*Also of note: there is no empty line at the end of this file.*

**Review by dog-m — CHANGES_REQUESTED** [review]

File: src/main/kotlin/org/jetbrains/research/libsl/visitors/ExpressionVisitor.kt (hunk 1 above)
Comment (alexeev509) [reply src/main/kotlin/org/jetbrains/research/libsl/visitors/ExpressionVisitor.kt:236]: Fixed )

File: src/main/antlr/LibSLLexer.g4 (hunk 2 above)
Comment (alexeev509) [reply src/main/antlr/LibSLLexer.g4]: Fixed )

File: src/main/kotlin/org/jetbrains/research/libsl/visitors/ExpressionVisitor.kt (hunk 4 above)
Comment (alexeev509) [reply src/main/kotlin/org/jetbrains/research/libsl/visitors/ExpressionVisitor.kt]: Fixed

File: src/test/testdata/lsl/charLiterals.lsl (hunk 7 above)
Comment (alexeev509) [reply src/test/testdata/lsl/charLiterals.lsl:7]: Was added

File: src/test/testdata/lsl/charLiterals.lsl (hunk 6 above)
Comment (alexeev509) [reply src/test/testdata/lsl/charLiterals.lsl]: Was deleted

File: src/test/testdata/expected/lsl/charLiterals.lsl (hunk 5 above)
Comment (alexeev509) [reply src/test/testdata/expected/lsl/charLiterals.lsl]: Was deleted

File: build.gradle.kts (hunk 3 above)
Comment (alexeev509) [reply build.gradle.kts]: It was added because when I have changed this implementation (in translator):

> implementation("com.github.vpa-research:libsl-parser:853ca7de04")


to this:

>     implementation(files("/Users/alekseevaroslav/Downloads/libsl-parser/build/libs/libsl-1.0-SNAPSHOT.jar"))


It doesn't find dependencies of the translator;

File: src/main/kotlin/org/jetbrains/research/libsl/visitors/ExpressionVisitor.kt (hunk 1 above)
Comment (dog-m) [reply src/main/kotlin/org/jetbrains/research/libsl/visitors/ExpressionVisitor.kt:236]: I think it is better to use [when expression](https://kotlinlang.org/docs/control-flow.html#when-expression) here.

File: src/main/kotlin/org/jetbrains/research/libsl/visitors/ExpressionVisitor.kt [hunk 9]
```diff
@@ -211,83 +224,86 @@ class ExpressionVisitor(
...
+        } else if (num.endsWith("l")) {
+            num = num.dropLast(1)
+            return IntegerLiteral(
+                convertBinHexOctToPrimitives(num, "l"),
+                "L",
+                posGetter.getCtxPosition(context.fileName, ctx)
+            )
         }
+        return IntegerLiteral(
+            convertBinHexOctToPrimitives(num, "i"),
```
Comment (dog-m) [inline src/main/kotlin/org/jetbrains/research/libsl/visitors/ExpressionVisitor.kt]: As I understand it, this implies that there are only `int32` and `int64` primitives can be expressed via 2-, 8- and 16-bit literals (binary, octal, hexadecimal).
Is this the intended behavior?

File: src/main/kotlin/org/jetbrains/research/libsl/visitors/ExpressionVisitor.kt [hunk 10]
```diff
@@ -211,83 +224,86 @@ class ExpressionVisitor(
...
+        val isInt = "i".equals(type)
+        val result: Number = if (num.startsWith(HEX_PREFIX)) {
+            if (isInt) parseInt(num.drop(2), 16) else parseLong(num.drop(2), 16)
+        } else if (num.startsWith(BIN_PREFIX)) {
+            if (isInt) parseInt(num.drop(2), 2) else parseLong(num.drop(2), 2)
+        } else if (num.startsWith(OCT_PREFIX) && num.length > 1) {
+            if (isInt) parseInt(num.drop(1), 8) else parseLong(num.drop(1), 8)
+        } else {
+            if (isInt) num.toInt() else num.toLong()
+        }
```
Comment (dog-m) [inline src/main/kotlin/org/jetbrains/research/libsl/visitors/ExpressionVisitor.kt]: May be something like this will be a bit better?
```kotlin
return = when {
    num.startsWith(HEX_PREFIX) -> isInt ? parseInt(num.drop(2), 16) : parseLong(num.drop(2), 16)
    num.startsWith(OCT_PREFIX) -> isInt ? parseInt(num.drop(2), 8) : parseLong(num.drop(2), 8)
    num.startsWith(BIN_PREFIX) -> isInt ? parseInt(num.drop(2), 2) : parseLong(num.drop(2), 2)
    _ -> isInt ? num.toInt() : num.toLong()
}
```

File: src/test/testdata/lsl/charLiterals.lsl (hunk 7 above)
Comment (dog-m) [reply src/test/testdata/lsl/charLiterals.lsl:7]: But what about non-printable characters? Like `new-line` and `caret-return`? `bell-character`? `null-character`?

File: src/test/testdata/expected/lsl/charLiterals.lsl [hunk 11]
```diff
@@ -0,0 +1,12 @@
+libsl "1.0.0";
+library simple;
+typealias Int = int32;
+typealias Char = char;
+
+automaton A : Int {
+    var ch1: Char = 'A';
+    var ch2: Char = 'a';
+    var ch3: Char = '狐';
```
Comment (dog-m) [inline src/test/testdata/expected/lsl/charLiterals.lsl]: Can this be read-back correctly by the parser? This symbol is not encoded as a Unicode code-point.

**Review by dog-m — CHANGES_REQUESTED** [review]

File: build.gradle.kts (hunk 3 above)
Comment (dog-m) [reply build.gradle.kts]: "Fat" libraries may cause problems: if there are multiple versions of such libraries are being used simultaneously, then there is a high chance they will be loaded incorrectly or conflict with each-other.

File: src/main/kotlin/org/jetbrains/research/libsl/visitors/ExpressionVisitor.kt (hunk 10 above)
Comment (alexeev509) [reply src/main/kotlin/org/jetbrains/research/libsl/visitors/ExpressionVisitor.kt]: Fixed

File: src/main/kotlin/org/jetbrains/research/libsl/visitors/ExpressionVisitor.kt (hunk 1 above)
Comment (alexeev509) [reply src/main/kotlin/org/jetbrains/research/libsl/visitors/ExpressionVisitor.kt:236]: Fixed

File: src/test/testdata/lsl/scientificNotation.lsl (hunk 8 above)
Comment (alexeev509) [reply src/test/testdata/lsl/scientificNotation.lsl]: Fixed;

File: src/test/testdata/expected/lsl/charLiterals.lsl (hunk 11 above)
Comment (alexeev509) [reply src/test/testdata/expected/lsl/charLiterals.lsl]: Fixed

File: src/main/kotlin/org/jetbrains/research/libsl/visitors/ExpressionVisitor.kt (hunk 9 above)
Comment (alexeev509) [reply src/main/kotlin/org/jetbrains/research/libsl/visitors/ExpressionVisitor.kt]: Yes, I suppose it will be enough... Or not ?

File: build.gradle.kts (hunk 3 above)
Comment (alexeev509) [reply build.gradle.kts]: Fixed

File: src/main/antlr/LibSLLexer.g4 [hunk 12]
```diff
@@ -254,6 +254,43 @@ NULL
...
+
+fragment HexIntegerLiteral: HexNumeral IntegerTypeSuffix?;
+
+fragment OctalIntegerLiteral: OctalNumeral IntegerTypeSuffix?;
+
+fragment BinaryIntegerLiteral: BinaryNumeral IntegerTypeSuffix?;
+
+fragment DecimalNumeral: '0' | NonZeroDigit (Digits?);
+
+fragment IntegerTypeSuffix: [lLbsu] | 'ub' | 'us' | 'uL';
```
Comment (dog-m) [inline src/main/antlr/LibSLLexer.g4]: To not have issues with `int8` literals I suggest switching to using `x` suffix instead of `b`.

Also, if it is possible to use `l` (small `L`) for `int64` literals, I strongly suggest adding similar case for *unsigned* literal variant (`ul`) too.

File: build.gradle.kts [hunk 13]
```diff
@@ -12,6 +12,7 @@ version = "1.0-SNAPSHOT"
 
 repositories {
     mavenCentral()
+    maven("https://jitpack.io")
```
Comment (dog-m) [inline build.gradle.kts]: Since "fat JAR"-related code was removed, does this line have a purpose?

File: src/main/antlr/LibSLLexer.g4 [hunk 14]
```diff
@@ -272,19 +309,42 @@ CHARACTER
...
+fragment ZeroToThree: [0-3];
+
+fragment Digits: Digit+;
+
 Digit: ('0'..'9');
 
-Hex: Digit | ('A'..'F');
+fragment NonZeroDigit: [1-9];
+
+fragment Hex: Digit | ('A'..'F');
```
Comment (dog-m) [inline src/main/antlr/LibSLLexer.g4]: Why not allow lower-case characters too?

File: src/main/antlr/LibSLLexer.g4 [hunk 15]
```diff
@@ -272,19 +309,42 @@ CHARACTER
...
+
+fragment HexNumeral: '0' [xX] Hex+;
```
Comment (dog-m) [inline src/main/antlr/LibSLLexer.g4:340]: Since the following prefix is expected to be at the beginning of a string, I consider forcing the user to write `0x` (lowercase)
```suggestion
fragment HexNumeral: '0x' Hex+;
```

..or convert to a lowercase/uppercase representation during parsing.

File: src/main/antlr/LibSLLexer.g4 [hunk 16]
```diff
@@ -272,19 +309,42 @@ CHARACTER
...
+
+fragment OctalNumeral: '0' OctalDigit+;
+
+fragment OctalDigit: [0-7];
+
+fragment BinaryNumeral: '0' [bB] BinaryDigit+;
```
Comment (dog-m) [inline src/main/antlr/LibSLLexer.g4:346]:
```suggestion
fragment BinaryNumeral: '0b' BinaryDigit+;
```
...or convert to lowercase/uppercase during parsing before checking the prefix.

File: src/test/testdata/lsl/charLiterals.lsl [hunk 17]
```diff
@@ -0,0 +1,13 @@
...
+    var ch2: Char = 'a';
+    var ch3: Char = '\u72D0';
+    var ch4: Char = '\52';
+    var ch5: Char = '\123';
+    var ch6: Char = '狐';
+}
```
Comment (dog-m) [inline src/test/testdata/lsl/charLiterals.lsl:25]: I suggest also adding a test case for non-printable characters. Like `\n`, `\r`, `\0` and such.

File: src/test/testdata/lsl/scientificNotation.lsl [hunk 18]
```diff
@@ -0,0 +1,13 @@
...
+typealias float = float32;
+typealias double = float64;
+automaton A : Int {
+    var d1: double = 4.05E-13;
+    var d2: double = 4.05E13;
+    var d3: double = 4.05E+13;
+    var d4: double = -1.5E+2;
+    var d5: double = -2.1E-5;
+    var d6: float = 3.4e+38f;
+}
```
Comment (dog-m) [inline src/test/testdata/lsl/scientificNotation.lsl:14]: I suggest also adding `0.0e+0` just to be extra sure.

File: src/main/kotlin/org/jetbrains/research/libsl/visitors/ExpressionVisitor.kt [hunk 19]
```diff
@@ -211,82 +224,81 @@ class ExpressionVisitor(
...
+            num.endsWith("b") -> IntegerLiteral(
+                ctx.text.dropLast(1).toByte(),
+                "b",
+                posGetter.getCtxPosition(context.fileName, ctx)
+            )
+            num.endsWith("us") -> UnsignedInt16Literal(
+                ctx.text.dropLast(2).toUShort(),
+                "us",
+                posGetter.getCtxPosition(context.fileName, ctx)
+            )
```
Comment (dog-m) [inline src/main/kotlin/org/jetbrains/research/libsl/visitors/ExpressionVisitor.kt]: Why not allow `unt16` hex/oct/bin literals (and others) too?

**Review by dog-m — CHANGES_REQUESTED** [review]

File: src/test/testdata/lsl/scientificNotation.lsl (hunk 18 above)
Comment (alexeev509) [reply src/test/testdata/lsl/scientificNotation.lsl:14]: Fixed;

File: src/main/antlr/LibSLLexer.g4 (hunk 14 above)
Comment (alexeev509) [reply src/main/antlr/LibSLLexer.g4]: fixed

File: src/main/antlr/LibSLLexer.g4 (hunk 15 above)
Comment (alexeev509) [reply src/main/antlr/LibSLLexer.g4:340]: I was looking at [this](https://github.com/antlr/grammars-v4/blob/master/java/java9/Java9Lexer.g4#L176) 
In class "ExpressionVisitor.kt" it converts to lowercase: 

```
    override fun visitIntegerNumber(ctx: LibSLParser.IntegerNumberContext): Atomic {
        var num = ctx.text.lowercase()
```

File: src/main/antlr/LibSLLexer.g4 (hunk 16 above)
Comment (alexeev509) [reply src/main/antlr/LibSLLexer.g4:346]: I was looking at [this](https://github.com/antlr/grammars-v4/blob/master/java/java9/Java9Lexer.g4#L176)
In class "ExpressionVisitor.kt" it converts to lowercase:

```
    override fun visitIntegerNumber(ctx: LibSLParser.IntegerNumberContext): Atomic {
        var num = ctx.text.lowercase()
```

File: src/main/kotlin/org/jetbrains/research/libsl/visitors/ExpressionVisitor.kt (hunk 19 above)
Comment (alexeev509) [reply src/main/kotlin/org/jetbrains/research/libsl/visitors/ExpressionVisitor.kt]: Fixed

File: src/main/antlr/LibSLLexer.g4 (hunk 12 above)
Comment (alexeev509) [reply src/main/antlr/LibSLLexer.g4]: Fixed;

File: src/test/testdata/lsl/charLiterals.lsl (hunk 17 above)
Comment (alexeev509) [reply src/test/testdata/lsl/charLiterals.lsl:25]: Fixed;

File: src/main/kotlin/org/jetbrains/research/libsl/nodes/helpers/ExpressionDumper.kt [hunk 20]
```diff
@@ -104,10 +104,24 @@ object ExpressionDumper {
...
+            '\n' -> "\\n"
+            '\r' -> "\\r"
+            '\t' -> "\\t"
+            '\b' -> "\\b"
+            '\'' -> "\\'"
+            '\"' -> "\\\""
+            '\\' -> "\\\\"
+            '\u0000' -> "\\0"
+            else -> String(value.toString().toByteArray(Charsets.UTF_8))
+        }
```
Comment (dog-m) [inline src/main/kotlin/org/jetbrains/research/libsl/nodes/helpers/ExpressionDumper.kt]: I suggest implementing a bit more standardized approach here:
* Use "pretty" variants for a specific set of characters.
* Convert to Unicode code points (like `\u1234`) characters that are either control characters or outside of the "Basic Latin" group of symbols.
* Print "Basic Latin" characters as is.

File: src/main/antlr/LibSLLexer.g4 [hunk 21]
```diff
@@ -272,19 +309,43 @@ CHARACTER
...
-SingleCharacter
+fragment SingleCharacter
     :   ~['\\\r\n]
     ;
 
-fragment
-EscapeSequence
-    :   '\\u' Hex Hex Hex Hex Hex Hex Hex Hex
+fragment EscapeSequence
+    :   '\\' 'u005c'? [btnfr"'\\]
```
Comment (dog-m) [inline src/main/antlr/LibSLLexer.g4]: What is the purpose of this expression?
What is `\u005cb` ? Why do we need to explicitly support it?

**Review by dog-m — CHANGES_REQUESTED** [review]

File: src/main/kotlin/org/jetbrains/research/libsl/nodes/helpers/ExpressionDumper.kt (hunk 20 above)
Comment (alexeev509) [reply src/main/kotlin/org/jetbrains/research/libsl/nodes/helpers/ExpressionDumper.kt]: Fixed;

File: src/main/antlr/LibSLLexer.g4 (hunk 21 above)
Comment (alexeev509) [reply src/main/antlr/LibSLLexer.g4]: Fixed;

File: build.gradle.kts (hunk 13 above)
Comment (alexeev509) [reply build.gradle.kts]: Fixed;

**Review by dog-m — APPROVED** [review]
